components/ucd/normal/tests/unicode_version_tests.rs

```diff
@@ -20,3 +21,149 @@ fn test_version_against_ucd_core() {
         unic_ucd_core::UNICODE_VERSION
     );
 }
+
+
+use unic_ucd_normal::DecompositionType as DT;
+use unic_ucd_utils::iter_all_chars;
+use std::collections::HashSet as Set;
+use std::{char, u32};
+
+
+const DT_TEST_DATA: &'static str =
+    include_str!("../../../../data/ucd/test/DecompositionTypeTest.txt");
+
+
+#[derive(Debug)]
+struct Fail {
+    pub line_num: Option<usize>,
+    pub char: char,
+    pub exp_dt: Option<DT>,
+    pub actual_dt: Option<DT>,
+}
+
+#[test]
+fn test_decomposition_type_conformance() {
+    let mut passed_num: u32 = 0;
+    let mut failures: Vec<Fail> = Vec::new();
+    let mut to_consider: Set<char> = iter_all_chars().collect();
+
+    for (line_idx, line) in DT_TEST_DATA.lines().enumerate() {
+        // Remove comments and trailing whitespace
+        let line = if let Some(idx) = line.find('#') {
+            &line[0..idx]
+        } else {
+            line
+        }.trim();
+
+        // Ignore empty lines
+        if line.is_empty() {
+            continue;
+        }
+
+        // State setting lines
+        if line.starts_with('@') {
+            panic!("State setting line in DecompositionTypeTest.txt; expected none");
+        }
+
+        // Data lines
+        {
+            let fields: Vec<_> = line.split(';').map(str::trim).collect();
+            let input_range = {
+                let range_text = fields[0];
+                if let Some(range_idx) = range_text.find("..") {
```

Comment:
Yeah, this turned out better, and seems more idiomatic with `Option` combinators.